### ISSUE TITLE
BM25 support

### ISSUE THREAD

**Issue by brightsparc** [quickwit-oss/tantivy#245]
I was wondering if there was any appetite for tantivy to support pluggable ranking functions, specifically i'm interested in implementing BM25.  
http://opensourceconnections.com/blog/2015/10/16/bm25-the-next-generation-of-lucene-relevation/
Lucene's doc is here: https://lucene.apache.org/core/7_0_1/core/org/apache/lucene/search/similarities/BM25Similarity.html

**Comment by fulmicoton** [comment]
This would be a nice addition. The current implementation of the tantivy's boolean scorer is extremely naive. If you feel like reimplementing it with BM25 that would be awesome. BM25 has the benefit of removing the need for "coords".

I am about to merge a less naive implementation in the case scoring is disabled. Maybe you can get some inspiration from it.

**Comment by fulmicoton** [comment]
The dependency is way too large to justify that, isn't it.
Having a similarity class like that is not a bad idea. 
With modern processors, dynamically taking this branch probably has no impact on perf.

The only possible downside I see is that it might get in the way when we implement WAND.

**Comment by brightsparc** [comment]
Yes fair enough, can just cherry pick things that make sense, will have a look at this once you have merged your current branch.

**Comment by fulmicoton** [comment]
I talked too fast. My change in the `BooleanScorer` should have no impact.
You can start working right now.

A bunch of pointer here :

- Fork from master and create a new branch called issue/245
- I believe the formula in kite is wrong. There are no sqrt in BM25.
- Feel free to introduce a `Similarity` enum, alternatively Similarity could be a generics of the TermScorer.
- Make sure to cache all multiplication that can be precomputed, and that scoring each document is fast.  You can rely array addressing too. You can check out how lucene does it
https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/similarities/BM25Similarity.java
- Your work will probably be contained in `TermWeight` and the `TermScorer`. 
In a nutshell Weight is the query applied to a specific set of segments.
In your case, you can precompute the total idf when you create the weight and store it in the termweight for instance. The `TermWeight`  creates the scorer, given a specific segment. That's where you want to 
precompute a bunch of values and store them in the `Scorer` object.
- Do NOT optimize the code that generates these precomputation. It will only be called once per segment.

**Comment by fulmicoton** [comment]
I added you as a developer. You can push your feature branch to the tantivy repo directly if you prefer.

Tantivy's gitter can help too https://gitter.im/tantivy-search/tantivy

**Comment by fulmicoton** [comment]
Stealing issue as there is no activity. Sorry @brightsparc

**Comment by brightsparc** [comment]
Sorry Paul, I’m been meaning to get to this. Still keen, will try and get to this in next month.

**Comment by fulmicoton** [comment]
It's okay. I think I will finish it today. Let me know if you want another ticket.
